src/modules/auth/auth.service.ts

```diff
@@ -1,43 +1,93 @@
-import { BadRequestException, Injectable } from '@nestjs/common';
+import { Hasher } from './../../shared/libs/hasher.lib';
+import {
+  BadRequestException,
+  ConflictException,
+  Injectable,
+  Logger,
+  UnauthorizedException,
+} from '@nestjs/common';
 import { JwtService } from '@nestjs/jwt';
 import { RegisterDto } from './dto/request/register.dto';
-import { Hasher } from '../../shared/libs/hasher.lib';
 import { UsersRepository } from '../users/users.repository';
 import { ConfigService } from '@nestjs/config';
 import { LoginDto } from './dto/request/login.dto';
 import { JwtPayload } from './auth.types';
 import { UpdatePasswordDto } from './dto/request/update-password.dto';
 import { User } from '../users/entities/user.entity';
+import { RedisService } from 'src/redis/redis.service';
+import { randomBytes } from 'node:crypto';
+import { MailerService } from '../mail/mail.service';
+import { VERIFICATION_LINK_EXPIRE_TIME } from 'src/shared/constants';
 
 @Injectable()
 export class AuthService {
+  private logger = new Logger(AuthService.name, { timestamp: true });
   constructor(
     private readonly usersRepository: UsersRepository,
     private readonly jwtService: JwtService,
     private readonly configService: ConfigService,
+    private readonly redisService: RedisService,
+    private readonly mailService: MailerService,
   ) {}
 
-  async register(registerDto: RegisterDto) {
+  private async prepareUserData(createUserDto: RegisterDto) {
+    const { email, password, firstName, lastName } = createUserDto;
+
     const existingUser = await this.usersRepository.findOne({
-      where: { email: registerDto.email },
+      where: { email },
     });
 
     if (existingUser) {
-      throw new BadRequestException('Email already exists in the system');
+      throw new BadRequestException('User with this email already exists');
     }
 
-    const hashedPassword = await Hasher.hashValue(registerDto.password);
+    const pendingKey = this.generateVerificationKey(email);
+    const existingPendingUser = await this.redisService.hGet(pendingKey, 'EX');
+    if (existingPendingUser) {
+      throw new ConflictException(
+        'Verification is already in progress for this email',
+      );
+    }
```

Comment:
Instead of throwing an error, you can send an ok reponse with the existing token saying "Verify your email" or "Verification was sent".

Reply:
But there will be a problem. the first time users get a verification email but don't look at it and try again, should I send an email again?